x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvPercentileTests.java

```diff
@@ -0,0 +1,433 @@
+
+                return new TestCaseSupplier.TestCase(
+                    List.of(fieldTypedData, percentileTypedData),
+                    evaluatorString(fieldSupplier.type(), percentileSupplier.type()),
+                    fieldSupplier.type(),
+                    expected instanceof Double expectedDouble
+                        ? closeTo(expectedDouble, Math.abs(expectedDouble * 0.0000001))
+                        : equalTo(expected)
+                );
+            }
+        );
+    }
+
+    private static Number calculatePercentile(List<Number> rawValues, double percentile) {
+        if (rawValues.isEmpty() || percentile < 0 || percentile > 100) {
+            return null;
+        }
+
+        if (rawValues.size() == 1) {
+            return rawValues.get(0);
+        }
+
+        int valueCount = rawValues.size();
+        var p = percentile / 100.0;
+        var index = p * (valueCount - 1);
+        var lowerIndex = (int) index;
+        var upperIndex = lowerIndex + 1;
+        var fraction = index - lowerIndex;
+
+        if (rawValues.get(0) instanceof Integer) {
+            var values = rawValues.stream().mapToInt(Number::intValue).sorted().toArray();
+
+            if (percentile == 0) {
+                return values[0];
+            } else if (percentile == 100) {
+                return values[valueCount - 1];
+            } else {
+                assert lowerIndex >= 0 && upperIndex < valueCount;
+                var difference = (long) values[upperIndex] - values[lowerIndex];
+                return values[lowerIndex] + (int) (fraction * difference);
+            }
+        }
+
+        if (rawValues.get(0) instanceof Long) {
+            var values = rawValues.stream().mapToLong(Number::longValue).sorted().toArray();
+
+            if (percentile == 0) {
+                return values[0];
+            } else if (percentile == 100) {
+                return values[valueCount - 1];
+            } else {
+                assert lowerIndex >= 0 && upperIndex < valueCount;
+                return calculatePercentile(fraction, new BigDecimal(values[lowerIndex]), new BigDecimal(values[upperIndex])).longValue();
+            }
+        }
+
+        if (rawValues.get(0) instanceof Double) {
+            var values = rawValues.stream().mapToDouble(Number::doubleValue).sorted().toArray();
+
+            if (percentile == 0) {
+                return values[0];
+            } else if (percentile == 100) {
+                return values[valueCount - 1];
+            } else {
+                assert lowerIndex >= 0 && upperIndex < valueCount;
+                return calculatePercentile(fraction, new BigDecimal(values[lowerIndex]), new BigDecimal(values[upperIndex])).doubleValue();
+            }
+        }
+
+        throw new IllegalArgumentException("Unsupported type: " + rawValues.get(0).getClass());
+    }
+
+    private static BigDecimal calculatePercentile(double fraction, BigDecimal lowerValue, BigDecimal upperValue) {
```

Comment:
As to avoid duplicating the exact logic there's in the original function, I'm always using BigDecimals gere to generate the cases (Except for ints, which is trivial).

Big chunk of code, but didn't find a better way